### PR TITLE
Semantics & Lowering for simple assignment.

### PR DESCRIPTION
**Pull request by fasiddique**

### CODE REVIEW

File: toolchain/semantics/semantics_node_kind.def [hunk 1]
```diff
@@ -28,6 +28,7 @@ CARBON_SEMANTICS_NODE_KIND(Invalid)
 CARBON_SEMANTICS_NODE_KIND(CrossReference)
 
 CARBON_SEMANTICS_NODE_KIND(Assign)
+CARBON_SEMANTICS_NODE_KIND(Assignment)
```
Comment (zygoloid) [inline toolchain/semantics/semantics_node_kind.def]: Can we use the same `Assign` node for both the variable declaration case and the assignment statement case? I think that was @jonmeow's intent here.

File: toolchain/semantics/semantics_handle_operator.cpp [hunk 2]
```diff
@@ -51,7 +51,14 @@ auto SemanticsHandleInfixOperator(SemanticsContext& context,
               resume_block_id));
       return true;
     }
-
+    case TokenKind::Equal: {
+      // TODO: handle complex assignment expression such as `a = a + 1`.
+      context.ImplicitAsRequired(
+          parse_node, lhs_id, context.semantics_ir().GetNode(rhs_id).type_id());
```
Comment (zygoloid) [inline toolchain/semantics/semantics_handle_operator.cpp]: I think we should do this the other way around: convert the RHS to the type of the LHS.

File: toolchain/semantics/semantics_handle_operator.cpp [hunk 3]
```diff
@@ -51,7 +51,14 @@ auto SemanticsHandleInfixOperator(SemanticsContext& context,
...
+      context.AddNodeAndPush(parse_node, SemanticsNode::Assignment::Make(
+                                             parse_node, lhs_id, rhs_id));
```
Comment (zygoloid) [inline toolchain/semantics/semantics_handle_operator.cpp]: We should check that the LHS is something that can be assigned to. I'm adding a mechanism to check that in #3055, so for now please just add a TODO here.

File: toolchain/lowering/testdata/operators/assignment.carbon [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+// CHECK:STDOUT: ; ModuleID = 'assignment.carbon'
+// CHECK:STDOUT: source_filename = "assignment.carbon"
+// CHECK:STDOUT:
+// CHECK:STDOUT: %TupleLiteralType = type { {}, {} }
+// CHECK:STDOUT: %TupleLiteralType.0 = type { i32, i32 }
+// CHECK:STDOUT:
+// CHECK:STDOUT: define void @Main() {
+// CHECK:STDOUT:   %var = alloca i32, align 4
+// CHECK:STDOUT:   store i32 12, ptr %var, align 4
+// CHECK:STDOUT:   store i32 -7, ptr %var, align 4
```
Comment (zygoloid) [inline toolchain/lowering/testdata/operators/assignment.carbon:15]: Why is this `-7` and not `9`?

Actually... this seems to be a pre-existing issue with the lowering of integer values; don't worry about it.

File: toolchain/semantics/semantics_handle_operator.cpp [hunk 5]
```diff
@@ -51,7 +51,15 @@ auto SemanticsHandleInfixOperator(SemanticsContext& context,
               resume_block_id));
       return true;
     }
-
+    case TokenKind::Equal: {
+      // TODO: handle complex assignment expression such as `a = a + 1`.
```
Comment (zygoloid) [inline toolchain/semantics/semantics_handle_operator.cpp]: I would expect `a = a + 1;` already works; did you mean `a += 1;` here?

Comment (fasiddique) [reply toolchain/semantics/semantics_handle_operator.cpp]: ah yes I meant to write a+=1. It's fixed now.

**Review by zygoloid — APPROVED** [review]

**Pull request merged**